frontend/catalyst/jax_primitives.py

```diff
@@ -366,6 +373,121 @@ def _print_lowering(jax_ctx: mlir.LoweringRuleContext, *args, string=None, memre
+
+transform_func_type = ir.OpaqueType.get("transform", "op<\"func.func\">", ir.Context())
+
+#
+# ???
+#
+class AbstractTransformFunc(AbstractValue):
+    """Abstract transform func type."""
+
+    hash_value = hash("AbstractTransformFunc")
+
+    def __eq__(self, other):
+        return isinstance(other, AbstractTransformFunc)
+
+    def __hash__(self):
+        return self.hash_value
+
+
+class ConcreteTransformFunc(AbstractTransformFunc):
+    """Concrete transform func type."""
+
+
+def _transform_func_lowering(aval):
+    assert isinstance(aval, AbstractTransformFunc)
+    return (transform_func_type,)
+
+core.raise_to_shaped_mappings[AbstractTransformFunc] = lambda aval, _: aval
+mlir.ir_type_handlers[AbstractTransformFunc] = _transform_func_lowering
+
+#
+# transform_named_sequence
+#
+@transform_named_sequence_p.def_abstract_eval
+def _transform_named_sequence_p_abstract_eval(*args):
+    return ()
+
+
+@transform_named_sequence_p.def_impl
+def _transform_named_sequence_p_def_impl(*args):  # pragma: no cover
+    raise NotImplementedError()
+
+
+def _transform_named_sequence_lowering(jax_ctx: mlir.LoweringRuleContext, *args):
+    functype = ir.FunctionType.get(inputs=[transform_func_type], results=[])
+    functype_attr = ir.TypeAttr.get(functype)
+
+    # add the transform.with_named_sequence unit attr to the parent module
+    module = jax_ctx.module_context.module
+    with_named_sequence_attr = ir.UnitAttr.get(jax_ctx.module_context.context)
+    module.operation.attributes["transform.with_named_sequence"] = with_named_sequence_attr
+
+    # Insert the transform.named_sequence op into the module to be transformed
+    with ir.InsertionPoint(module.body):
+        named_sequence_op = NamedSequenceOp(sym_name="__transform_main",
+        function_type=functype_attr, 
+        )
+
+        # transform.named_sequence op is the "main function" of the transform dialect
+        # and thus needs an entry block
+        # The argument of the block is the target function to run the transform sequence on
+        bb = ir.Block.create_at_start(named_sequence_op.body, 
+        arg_types = [transform_func_type])
+
+        # The transform.named_sequence needs a terminator called "transform.yield"
+        with ir.InsertionPoint(bb):
+            transform_yield_op = TransformYieldOp(operands_=[])
+
+    return named_sequence_op.results
+
+
+#
+# apply_registered_pass
+#
+@apply_registered_pass_p.def_abstract_eval
+def _apply_registered_pass_abstract_eval(*args, pass_name, options=None):
+    return AbstractTransformFunc()
+
+@apply_registered_pass_p.def_impl
+def _apply_registered_pass_def_impl(*args, pass_name, options=None):  # pragma: no cover
+    raise NotImplementedError()
+
+
+def _apply_registered_pass_lowering(jax_ctx: mlir.LoweringRuleContext, *args, pass_name, options=None):
+    module = jax_ctx.module_context.module
+    named_sequence_op = None
+    for i in range(len(module.body.operations)):
```

Comment:
TODO: iterate from the end instead of from the start, since the named_sequence is injected into the module at the end

Reply:
... and break after the `named_sequence` is found